### DISCUSSION TITLE
Person recognition mistake

### DISCUSSION THREAD

**Discussion by jdbaron25** [photoprism/photoprism#1926] [General]
I made a mistake and mislabeled a person.  Now, thousands of photos have the wrong person label.  Is there a way to get rid of all person labels and start over?

**Comment by acranox** [comment]
https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface
**Reset People & Faces** `docker-compose exec photoprism photoprism faces reset -f`

**Reply by nautatva** [reply]
Will this reset all the faces? Option to change name for only one face?

**Reply by graciousgrey** [reply]
The command will reset everything. 

If a face is matched to the wrong person you can change this using the web ui: https://docs.photoprism.app/user-guide/organize/people/#change-people-assignments.

If you want to rename a person (this will affect all photos of this person) do this: https://docs.photoprism.app/user-guide/organize/people/#rename-people

**Reply by nautatva** [reply]
When saying `Reset People & Faces`, Will the recognition models be reset as well? (I assume the software is learning when I retag my image, not sure on this point).

Another thing I couldn't find in documentation: Complete Rescan will keep the faces and just reattach the images? Or is it a complete reset?

**Reply by lastzero** [reply]
You can find a short description of how facial recognition works here: https://docs.photoprism.app/getting-started/config-options/#facial-recognition

**Reply by nautatva** [reply]
Oh no, red alert, noob user here, the workings aren't helping me find a answer. I will reiterate the question.
I have already labelled some of my faces, now if I do a complete rescan, will the labelling be lost? Same question, if i do **Reset People & Faces** `docker-compose exec photoprism photoprism faces reset -f`, will the things learned because of some corrections during labelling be lost?

**Reply by graciousgrey** [reply]
The faces reset command will reset everything face related, so that you can start from scratch.
A complete re-index does update the index but does not overwrite manual changes such as labelled faces.

**Reply by OmegaRa** [reply]
> https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface **Reset People & Faces** `docker-compose exec photoprism photoprism faces reset -f`

I would like to do this, but I am running PhotoPrism as an app in CasaOS and don't know where to use this command, I tried in terminal (Ubunutserver) but it didn't recognize it

**Comment by lapo-luchini** [comment]
I have have labeled ~200 people correctly, but I mis-clicked one photo of (John A) as (John B)… it seems that now all the "face clusters" of John A & B have been merged: if I "eject" a photo of John A, and re-label it as John B then all photos of John A (even the ones that are correctly John A) get labeled as John B.

Resetting al faces is not a solution as I'd have to re-label all ~200 people… is there any way to say "I did an error on this single group of people, just delete this cluster and re0build it on next re-index"?

**Reply by graciousgrey** [reply]
Not yet. Currently there is only a command to reset all faces.

**Reply by lapo-luchini** [reply]
I'll wait for the feature in the future then, I prefer to have two merged people than re-tagging them all. =)

**Reply by lapo-luchini** [reply]
PS: I have no problems using SQL to reset a single face if that's possible.

**Reply by ronhaley48** [reply]
Same issue.  Not willing to throw away hours of labelling to fix one or two people. But when the person is your wife and her identical twin, it's a bit of an issue:)

**Reply by cincodenada** [reply]
I had this issue and know my way around a database and was able to reassign the erroneous face, and figured I'd share my success here for future searchers.

DISCLAIMER: if you don't know what you're doing, don't do it, you can irrecoverably break a lot of stuff with a bad SQL query, and even this modification is "out of warranty" - I don't know if I missed something that will make things wonky down the road. I'm all for learning how to do stuff, but you don't want to learn the hard way by hosing your photo collection. If you want to learn, back up your database first (and learn how to do that first if you don't know how!)

So that said, there are two tables that I updated: `faces`, which maps people ("subjects") to recognized faces, and `markers`, which maps recognized faces to files.

First you'll need to figure out what face ID you want to reassign. I did this by finding a photo with the misassigned face and listing the faces on that photo. You can get the photo UID from the farthest right :gear: tab on the photo (note: not the "Files" tab, that's a different UID).

Once I had that, I ran this query:
```
SELECT f.id, f.file_name, m.subj_uid, s.subj_name, m.face_id FROM photos p LEFT JOIN files f ON f.photo_id=p.id LEFT JOIN markers m ON m.file_uid=f.file_uid LEFT JOIN subjects s ON s.subj_uid=m.subj_uid WHERE p.photo_uid=LOWER("<YOUR PHOTO UID HERE>");
```
This will give you a list of face id's present in that photo, with the subjects/names currently assigned to them. Grab the `face_id` of the face that was misassigned - you can confirm by searching for `face:<YOUR FACE ID>` in photoprism to show the photos with the given face.

Now you need to find the correct subject ID to assign it to - this should be as easy as `SELECT subj_uid, subj_name FROM subjects;` and finding the right subject, if you have a _lot_ of subjects you can filter with LIKE or what have you.

Okay, once you have the face ID and correct subject ID you need to remap them in the two tables. This is the dangerous part! Turn back here if you are uncertain, make a backup of your database to be extra safe, etc etc.

First, check the existing face record:
```
SELECT id, subj_uid FROM faces f WHERE f.id="<YOUR FACE ID HERE>";
```
There should be one row, this is what links this face to the given subject. It should have a different (wrong) `subj_uid` currently. If all looks good, update it to the correct id:
```
UPDATE faces SET subj_uid="<CORRECT SUBJECT ID HERE>" WHERE id="<YOUR FACE ID HERE>" LIMIT 1;
```
(The `LIMIT 1` is not strictly necessary, but it's a good habit, as it limits the damage if you screw something up)

If you go and check now, nothing will be fixed, because this match is propagated onto each file with the `markers` table. Again, do a pre-check and query that table:
```
SELECT file_uid, subj_uid, subj_src, face_id FROM markers WHERE face_id="<YOUR FACE ID HERE>";
```
This will list all the files with the given face ID, likely one will have "manual" in the `subj_src` field (this is the one you tagged). Make sure this list looks reasonable, and then update the mappings:
```
UPDATE markers SET subj_uid="<CORRECT SUBJECT ID HERE>" WHERE face_id="<YOUR FACE ID HERE>";
```
If all went well, you should be able to refresh your library and see the faces correctly mapped.

The only inconsistency I noticed was that the count of the affected photos in the People view was out of date, but it updated the next time I tagged a photo, as expected. Good luck!
